### ISSUE TITLE
Options documentation is confusing.

### ISSUE THREAD

**Issue by AlansCodeLog** [metadevpro/ts-pegjs#44]
First in the readme some options are named with dashes and others aren't, not sure why?

Also I was trying to get `returnTypes` working and since there's no examples and it was grouped with the plugin options that went under the `tspegjs` property, I assumed it would go there. Had to search the code to find out what was going on.

A example in the readme using all the options possible would be nice.

Additionally I can't get returnTypes to work. I mean I can see it's typing all the rule functions, but since `peg$parsestart` is wrapped by `peg$startRuleFunction` and  `peg$parse` is further typed as `ParseFunction`, both returning any, even if I set the start rule type, the type of parse is always any. Is this intended? Should I open an issue?

**Comment by Enteee** [comment]
Yep can confirm. Just spent a few hours trying to get `returnType` working before checking the code. 

> Additionally I can't get returnTypes to work. I mean I can see it's typing all the rule functions, but since peg$parsestart is wrapped by peg$startRuleFunction and peg$parse is further typed as ParseFunction, both returning any, even if I set the start rule type, the type of parse is always any. Is this intended? Should I open an issue?

There is more information about this in this issue: #11 . Especially in [this comment by @pjmolina](https://github.com/metadevpro/ts-pegjs/issues/11#issuecomment-392247885). If I get this right, @pjmolina  suggest wrapping `parse`.

I am struggling a bit with typing intermediate production rules. Which always seem to fail. A bit of documentation around this would be nice as well.

**Comment by wy193777** [comment]
So if I understand it right, there is no way to specify types for `parse` function from `returnTypes`?

**Comment by Enteee** [comment]
@wy193777 as far as I understand it this is correct. But you can do what I did and [wrap `parse`](https://github.com/Enteee/plantuml-parser/blob/master/src/index.ts#L33) to force a typed `parse`.

**Comment by pjmolina** [comment]
> @AlansCodeLog First in the readme some options are named with dashes and others aren't, not sure why?

- Options with long names uses double-dashes `--` like in `--cache`.
- Short-hands options (one letter) uses single-dashes `-` like in `-o`, some time they can be grouped when working as binary flags like in `-oc`.
- In code options (JS) uses no dashes as it is JSON objects.
The first two conventions are inherited from `pegjs` and this inherits for common practice in the UNIX/POSIX world.

> A example in the readme using all the options possible would be nice.

You are right, I will prepare a better documentation for this.
So far, the best example is to take a look to [one example here](https://github.com/metadevpro/ts-pegjs/blob/dc769c4771693fe6e03a14e78be87323e89ca558/test/test.js#L45). 

As commented, ts-pegjs is a plugin for pegjs this involves:
- knowledge about how pegjs works (not repeating here what is already documented on pegjs)
- it is a soft rewrite to TS of what is actually generated by JS.
- so typing here is more than a hack on the final public production rules than a proper and complete typings. As discussed [before](https://github.com/pegjs/pegjs/issues/562) complete typings should be done extending the pegjs grammar to accept types as first citizens.

**Comment by AlansCodeLog** [comment]
> Options with long names uses double-dashes -- like in --cache.
Sorry I wasn't clearer, I meant that some of the options are named like `some-option` and some in `CamelCase`. Felt a bit weird and inconsistent.

Also I tested the example, and it has the same issue I described. The export ends up getting typed as any. I filed the issue because I was wondering if this was intentional. I mean, I guess it's a bit helpful for the output to be typed, but I was expecting the plugin to type the `parse` function it exports (using the start rule, or a union of possible rules) for *consumption*. I have ended up just re-exporting the function with the types I want, but it would help to clarify this needs to be done.

> so typing here is more than a hack on the final public production rules than a proper and complete typings. As discussed before complete typings should be done extending the pegjs grammar to accept types as first citizens.

Yeah too bad development seems to have stalled. Anyways, I have something satisfactory working, and at least now there's this issue if others are wondering about this.
